### DISCUSSION TITLE
Prisma Client in AWS Lambda with TypeScript: "ENOENT: no such file or directory, open '/var/task/src/functions/users/schema.prisma'"

### DISCUSSION THREAD

**Discussion by nikolasburk** [prisma/prisma#15697] [Q&A] [answered]
I'm trying to integrate Prisma into the basic [`aws-node-typescript`](https://github.com/serverless/examples/tree/v3/aws-node-typescript) example.

So far I have:

1. Created the project locally using `serverless create`
2. Set up a PostgreSQL database on Railway
3. Installed `prisma`, ran `prisma init`, created a basic `User` model and ran `prisma migrate dev` successfully
4. Deployed the function using `serverless deploy`
5. Now in my function, when I instantiate `PrismaClient`, I get an internal server error and the function logs this error: `"ENOENT: no such file or directory, open '/var/task/src/functions/users/schema.prisma'"`

For reference, here's my Prisma schema:

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id    Int     @id @default(autoincrement())
  name  String?
  email String  @unique
}
```

Here's my handler: 

```ts
import { formatJSONResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

const users = async (event) => {
  return formatJSONResponse({
    message: `Hello, ${event.queryStringParameters.name || 'there'} welcome to the exciting Serverless world!`,
    event,
  });
};

export const main = middyfy(users);
```

As you can see, I'm not eve at the point where I would use the `prisma` instance inside the handler, but just instantiating `PrismaClient` already creates this error.

**Comment by austincrim** [comment]
I think these kinds of errors come down to the bundling/packaging configuration. Prisma requires the `schema.prisma` file to be present during runtime, and it seems the default Serverless packaging configuration does not include the necessary files/directories.

If you run `serverless package`, you can look inside the `.serverless` folder to see what artifacts will actually be deployed with the function. You can then configure the [`serverless.yml` `package.patterns` field](https://www.serverless.com/framework/docs/providers/aws/guide/packaging#package-configuration) to get the right stuff in the deployed function.

**Comment by nikolasburk** [comment, marked as answer]
For anyone finding this, I wrote up a solution to the problem on Stackoverflow: https://stackoverflow.com/questions/73985623/package-files-into-specific-folder-of-application-bundle-when-deploying-to-aws-l

**Comment by jeanhdev** [comment]
For anyone struggling with this, the only solution I found is to NOT use `serverless-esbuild` and instead use `serverless-plugin-typescript`... Longer builds but works like a charm with Prisma.

**Comment by andenacitelli** [comment]
I ran into this in a `pnpm` monorepo setup. Here's my workaround:

1. Manually copy `schema.prisma` from original location into functions folder
2. Manually copy any `libquery_engine-rhel*` file from original location into functions folder
3. Configure `serverless.yml` to only bundle the specific binary you're looking for, which you can switch by `stage`

I was unable to find a solution that wouldn't physically copy and duplicate those files in, but as long as you make sure they're covered by your `.gitignore` and you automate it to some extent with scripts, I don't see it as a huge deal.

Relevant File Structure:
```
/ (monorepo root)
  /apps
    /server
  /lib
    /db
```

Relevant portions of `schema.prisma` in db package:

```
datasource db {
  provider = "mysql"
  url      = "mysql://..."
}

generator client {
  provider = "prisma-client-js"
  output   = "../src/generated/client"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
}

model Repository {
  id          Int      @id @default(autoincrement())
 // ...
}
```

Important part is the `binaryTargets` line.

Server package's `deploy:prod` command: 

```shell
"cp -r ../../lib/db/src/generated/client/{schema.prisma,libquery_engine-rhel*} src && sls deploy --stage prod
```

Relevant parts of server package's `serverless.yml`:

```
custom:
  stage: ${opt:stage, 'offline'}
  region: ${opt:region, 'us-east-1'}
  prismaEngine:
    offline: "src/*.node"
    staging: "src/libquery_engine-rhel*"
    prod: "src/libquery_engine-rhel*"
  esbuild:
    platform: "node"
    target: node18
    sourcemap: linked

package:
  patterns:
    - "!**/*.test.ts"
    - ${self:custom.prismaEngine.${self:custom.stage}}
    - "src/schema.prisma"
  individually: true

plugins:
  - serverless-esbuild
```

Key part of the `serverless.yml` is how we dynamically determine which binary we're looking for based off of the stage that was passed in. If we're on prod, we'll copy the version of the binary that works with Lambda. If we're not in prod, we'll copy the native version.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
